### ISSUE TITLE
Order of rules executed

### ISSUE THREAD

**Issue by SefaOray** [JeremySkinner/FluentValidation#699]
I have a validator like this

 ```
public ProductValidator()
{
    When(p => p.BuyPrice!= null, BuyPriceValidations);
}

private void BuyPriceValidations()
{
    RuleFor(p => p.BuyPrice.Value).Cascade(CascadeMode.StopOnFirstFailure)

    .NotNull()

    .NotEmpty().WithMessage("Buy price can not be empty.").WithErrorCode("BuyPriceEmpty")

     .GreaterThan(0).WithMessage("Buy price must be greater than 0.").WithErrorCode("BuyPriceEmpty")

    .GreaterThan(p => p.SaleBuyPrice.Value)
    .When(p => p.SaleBuyPrice != null)
    .WithMessage("Buy price must be greater than sale buy price")

}
```

I wrote a unit test to test first two `NotNull` and `NotEmpty` rules. 

```
[TestMethod]
public void ShoulHaveErrorWhenBuyPriceIsNull()
{
    var product = new Product() { BuyPrice = new ProductProperty<decimal>() };
    var result = validator.ShouldHaveValidationErrorFor(p => p.BuyPrice.Value, product);
    result.WithErrorCode("BuyPriceEmpty");
}
```

The problem is product is passing `NotNull` and `NotEmpty` rules and failing on `Buy price must be greater than sale buy price` rule. The test is throwing the following exception:

`FluentValidation.TestHelper.ValidationTestException: Expected an error code of 'BuyPriceEmpty'.`

Since I am not giving BuyPrice any value, it's 0, as decimals default value. 

It seems to me that last rule is executed first. Is there a way to order my rules or am I designing the whole thing wrong?

**Comment by JeremySkinner** [comment]
Hi,

> It seems to me that last rule is executed first.

No that's not the case. Rules are always executed in the order they've been defined (unless you're using async validation where the order isn't guaranteed). 

It looks to me that your rule might not be quite right. You have a call to `When(p => p.SaleBuyPrice != null)` - When conditions are applied to *all preceding validators in the chain* (including the NotNull/NotEmpty), and because your test does not set SaleBuyPrice, none of the rules will be executed. 

I'm guessing that you only want that condition to apply to the `.GreaterThan(p => p.SaleBuyPrice.Value)`? If so then you have two options. Firstly, there's a second overload of the `When` method that allows you to specify this, but personally I'd split them into two separate rules as it makes it much easier to read and understand.

Hope that helps.

**Comment by SefaOray** [comment]
Hi,

It does. Thanks a lot!
